File: src/k6s/models/context.py
"""Context storage model for persistent agent context."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class ContextEntry(BaseModel):
    """Represents a key-value context entry that persists across sessions."""

    key: str
    session_id: str
    agent_id: str | None = None
    value: Any
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    def to_db_row(self) -> dict[str, Any]:
        """Convert to database row format."""
        import json

        return {
            "key": self.key,
            "session_id": self.session_id,
            "agent_id": self.agent_id,
            "value": json.dumps(self.value),
            "updated_at": self.updated_at.isoformat(),
        }

    @classmethod
    def from_db_row(cls, row: dict[str, Any]) -> "ContextEntry":
        """Create context entry from database row."""
        import json

        return cls(
            key=row["key"],
            session_id=row["session_id"],
            agent_id=row["agent_id"],
            value=json.loads(row["value"]),
            updated_at=datetime.fromisoformat(row["updated_at"]),
        )


class FileLock(BaseModel):
    """Represents a file lock held by an agent."""

    path: str
    session_id: str
    agent_id: str
    acquired_at: datetime = Field(default_factory=datetime.utcnow)
    expires_at: datetime | None = None

    @property
    def is_expired(self) -> bool:
        """Check if lock has expired."""
        if self.expires_at is None:
            return False
        return datetime.utcnow() > self.expires_at

    def to_db_row(self) -> dict[str, Any]:
        """Convert to database row format."""
        return {
            "path": self.path,
            "session_id": self.session_id,
            "agent_id": self.agent_id,
            "acquired_at": self.acquired_at.isoformat(),
            "expires_at": self.expires_at.isoformat() if self.expires_at else None,
        }

    @classmethod
    def from_db_row(cls, row: dict[str, Any]) -> "FileLock":
        """Create file lock from database row."""
        return cls(
            path=row["path"],
            session_id=row["session_id"],
            agent_id=row["agent_id"],
            acquired_at=datetime.fromisoformat(row["acquired_at"]),
            expires_at=datetime.fromisoformat(row["expires_at"]) if row["expires_at"] else None,
        )


class BoundaryViolation(BaseModel):
    """Represents a boundary violation by an agent."""

    id: str
    session_id: str
    agent_id: str | None = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    file_path: str
    violation_type: str  # "forbidden_path" | "outside_allowed" | "resource_limit"
    enforcement_action: str  # "logged" | "reverted" | "blocked"
    details: dict[str, Any] = Field(default_factory=dict)

    def to_db_row(self) -> dict[str, Any]:
        """Convert to database row format."""
        import json

        return {
            "id": self.id,
            "session_id": self.session_id,
            "agent_id": self.agent_id,
            "timestamp": self.timestamp.isoformat(),
            "file_path": self.file_path,
            "violation_type": self.violation_type,
            "enforcement_action": self.enforcement_action,
            "details": json.dumps(self.details) if self.details else None,
        }

    @classmethod
    def from_db_row(cls, row: dict[str, Any]) -> "BoundaryViolation":
        """Create violation from database row."""
        import json

        return cls(
            id=row["id"],
            session_id=row["session_id"],
            agent_id=row["agent_id"],
            timestamp=datetime.fromisoformat(row["timestamp"]),
            file_path=row["file_path"],
            violation_type=row["violation_type"],
            enforcement_action=row["enforcement_action"],
            details=json.loads(row["details"]) if row["details"] else {},
        )

File: src/models/context.ts
/**
 * Context storage, file lock, and boundary violation models.
 */

import { z } from "zod";
import type { Row } from "../store/db.js";

// Context entry

export const ContextEntrySchema = z.object({
  key: z.string(),
  sessionId: z.string(),
  agentId: z.string().nullable().default(null),
  value: z.string(),
  updatedAt: z.string().default(() => new Date().toISOString()),
});
export type ContextEntry = z.infer<typeof ContextEntrySchema>;

export function contextEntryToDbRow(e: ContextEntry): Row {
  return {
    key: e.key,
    session_id: e.sessionId,
    agent_id: e.agentId,
    value: e.value,
    updated_at: e.updatedAt,
  };
}

export function contextEntryFromDbRow(row: Row): ContextEntry {
  return {
    key: row.key as string,
    sessionId: row.session_id as string,
    agentId: (row.agent_id as string) ?? null,
    value: row.value as string,
    updatedAt: row.updated_at as string,
  };
}

// File lock

export const FileLockSchema = z.object({
  path: z.string(),
  sessionId: z.string(),
  agentId: z.string(),
  acquiredAt: z.string().default(() => new Date().toISOString()),
  expiresAt: z.string().nullable().default(null),
});
export type FileLock = z.infer<typeof FileLockSchema>;

export function fileLockToDbRow(l: FileLock): Row {
  return {
    path: l.path,
    session_id: l.sessionId,
    agent_id: l.agentId,
    acquired_at: l.acquiredAt,
    expires_at: l.expiresAt,
  };
}

export function fileLockFromDbRow(row: Row): FileLock {
  return {
    path: row.path as string,
    sessionId: row.session_id as string,
    agentId: row.agent_id as string,
    acquiredAt: row.acquired_at as string,
    expiresAt: (row.expires_at as string) ?? null,
  };
}

export function isLockExpired(l: FileLock): boolean {
  if (!l.expiresAt) return false;
  return new Date() > new Date(l.expiresAt);
}

// Boundary violation

export const BoundaryViolationSchema = z.object({
  id: z.string(),
  sessionId: z.string(),
  agentId: z.string().nullable().default(null),
  timestamp: z.string().default(() => new Date().toISOString()),
  filePath: z.string(),
  violationType: z.string(),
  enforcementAction: z.string(),
  details: z.string().nullable().default(null),
});
export type BoundaryViolation = z.infer<typeof BoundaryViolationSchema>;

export function boundaryViolationToDbRow(v: BoundaryViolation): Row {
  return {
    id: v.id,
    session_id: v.sessionId,
    agent_id: v.agentId,
    timestamp: v.timestamp,
    file_path: v.filePath,
    violation_type: v.violationType,
    enforcement_action: v.enforcementAction,
    details: v.details,
  };
}

export function boundaryViolationFromDbRow(row: Row): BoundaryViolation {
  return {
    id: row.id as string,
    sessionId: row.session_id as string,
    agentId: (row.agent_id as string) ?? null,
    timestamp: row.timestamp as string,
    filePath: row.file_path as string,
    violationType: row.violation_type as string,
    enforcementAction: row.enforcement_action as string,
    details: (row.details as string) ?? null,
  };
}
